### PR TITLE
Seg 8

### PR DESCRIPTION
**Pull request by vjnathe-webonise**
Added test cases for Facebook Pixel Standard Events.
We have added missing properties in mapping for 4 standard events for this we have created mapping sheet, which basically maps the segment specs to facebook pixel document and  we added test case for those standard events for which we found difference.

### CODE REVIEW

File: integrations/facebook-pixel/test/index.test.js [hunk 1]
```diff
@@ -1344,6 +1343,33 @@ describe('Facebook Pixel', function() {
         );
         assertEventId(window.fbq);
       });
+
+      it('should send facebook pixel properties for product viewed', function() {
+        analytics.track('Product Viewed', {
+          product_id: '507f1f77bcf86cd799439011',
+          contents: [
```
Comment (brennan) [inline integrations/facebook-pixel/test/index.test.js]: A `contents` array shouldn't appear in a Segment track event, but the logic should maps products from a Segment event to a Facebook `contents` array. Example: https://github.com/webonise/analytics.js-integrations/blob/master/integrations/facebook-pixel/test/index.test.js#L2047-L2050

File: integrations/facebook-pixel/test/index.test.js [hunk 2]
```diff
@@ -1721,6 +1747,33 @@ describe('Facebook Pixel', function() {
         });
         assertEventId(window.fbq);
       });
+
+      it('should send facebook pixel properties for product added', function() {
+        analytics.track('Product Added', {
+          product_id: '507f1f77bcf86cd799439011',
+          contents: [
```
Comment (brennan) [inline integrations/facebook-pixel/test/index.test.js]: Same here - a `contents` array doesn't appear in the Segment spec.

File: integrations/facebook-pixel/test/index.test.js [hunk 3]
```diff
@@ -1942,6 +1995,51 @@ describe('Facebook Pixel', function() {
...
+          'Order Completed',
+          {
+            products: [
+              { product_id: '507f1f77bcf86cd799439011', category: 'Cars' },
+              { product_id: '505bd76785ebb509fc183733', category: 'Cars' }
+            ],
+            currency: 'USD',
+            total: 0.5,
+            content_name: 'my product',
+            contents: [
```
Comment (brennan) [inline integrations/facebook-pixel/test/index.test.js]: There shouldn't be a `contents` array in a Segment track event.

File: integrations/facebook-pixel/test/index.test.js [hunk 4]
```diff
@@ -1942,6 +1995,51 @@ describe('Facebook Pixel', function() {
...
+        analytics.track(
+          'Order Completed',
+          {
+            products: [
+              { product_id: '507f1f77bcf86cd799439011', category: 'Cars' },
+              { product_id: '505bd76785ebb509fc183733', category: 'Cars' }
+            ],
+            currency: 'USD',
+            total: 0.5,
+            content_name: 'my product',
```
Comment (brennan) [inline integrations/facebook-pixel/test/index.test.js]: `content_name` is not a Segment-specced property.

File: integrations/facebook-pixel/test/index.test.js [hunk 5]
```diff
@@ -1942,6 +1995,51 @@ describe('Facebook Pixel', function() {
...
+            contents: [
+              { id: '507f1f77bcf86cd799439011', quantity: 1, item_price: 19 }
+            ],
+            num_items: 2
```
Comment (brennan) [inline integrations/facebook-pixel/test/index.test.js]: `num_items` is not a Segment-specced property.

File: integrations/facebook-pixel/test/index.test.js [hunk 6]
```diff
@@ -1992,6 +2090,30 @@ describe('Facebook Pixel', function() {
 
         analytics.called(window.fbq, 'trackSingle', options.pixelId, '123456');
       });
+
+      it('should send facebook pixel properties for product search', function() {
+        analytics.track('Products Searched', {
+          query: 'yo',
+          content_category: 'Cars',
```
Comment (brennan) [inline integrations/facebook-pixel/test/index.test.js]: `content_category`, `content_ids`, and `contents` are not Segment-specced properties, so shouldn't appear in a `track` call.

File: integrations/facebook-pixel/test/index.test.js [hunk 7]
```diff
@@ -1992,6 +2090,30 @@ describe('Facebook Pixel', function() {
...
+          content_ids: ['507f1f77bcf86cd799439011'],
+          contents: [
+            { id: '507f1f77bcf86cd799439011', quantity: 1, item_price: 19 }
+          ],
+          currency: 'USD',
+          value: 30
```
Comment (brennan) [inline integrations/facebook-pixel/test/index.test.js]: `value` isn't a Segment-specced property, so shouldn't appear here.

**Review by brennan — CHANGES_REQUESTED** [review]
Looks like there may be confusion about Segment-specced properties/how properties are mapped from Segment -> Facebook. Two next steps:
- Please share the mapping sheet @vjnathe-webonise created.
- Please review some of the existing tests/logic and let me know what questions you have about mapping. Existing mappings should be clear, such as how Segment maps `category` to Facebook `content_category` (this is just one example of many): https://github.com/webonise/analytics.js-integrations/blob/master/integrations/facebook-pixel/lib/index.js#L273

File: integrations/facebook-pixel/test/index.test.js [hunk 8]
```diff
@@ -1344,6 +1343,38 @@ describe('Facebook Pixel', function() {
         );
         assertEventId(window.fbq);
       });
+
+      it('should send facebook pixel properties for product viewed', function() {
+        analytics.track('Product Viewed', {
```
Comment (brennan) [inline integrations/facebook-pixel/test/index.test.js:1405]: "Product Viewed" events are formatted a little differently than what you have here: https://segment.com/docs/spec/ecommerce/v2/#product-viewed

File: integrations/facebook-pixel/test/index.test.js [hunk 9]
```diff
@@ -1344,6 +1343,38 @@ describe('Facebook Pixel', function() {
...
+        });
+        analytics.called(
+          window.fbq,
+          'trackSingle',
+          options.pixelId,
+          'ViewContent',
+          {
+            content_ids: ['507f1f77bcf86cd799439011'],
+            content_type: ['product'],
+            content_name: '',
```
Comment (brennan) [inline integrations/facebook-pixel/test/index.test.js]: This test wouldn't pass after we implement the updated logic correctly. After updating the logic, we would map `name` in the track event to `content_name` in the Facebook Standard event. So, the value of `content_name` here should be "'Monopoly: 3rd Edition'".

File: integrations/facebook-pixel/test/index.test.js [hunk 10]
```diff
@@ -1721,6 +1752,38 @@ describe('Facebook Pixel', function() {
         });
         assertEventId(window.fbq);
       });
+
+      it('should send facebook pixel properties for product added', function() {
+        analytics.track('Product Added', {
```
Comment (brennan) [inline integrations/facebook-pixel/test/index.test.js:1842]: "Product Added" events are formatted like this: https://segment.com/docs/spec/ecommerce/v2/#product-added

File: integrations/facebook-pixel/test/index.test.js [hunk 11]
```diff
@@ -1721,6 +1752,38 @@ describe('Facebook Pixel', function() {
...
+          'trackSingle',
+          options.pixelId,
+          'AddToCart',
+          {
+            content_ids: ['507f1f77bcf86cd799439011'],
+            content_type: ['product'],
+            contents: [
+              { id: '507f1f77bcf86cd799439011', quantity: 1, item_price: 19 }
+            ],
+            content_name: '',
```
Comment (brennan) [inline integrations/facebook-pixel/test/index.test.js]: Same content as above re: "content_name".

File: integrations/facebook-pixel/test/index.test.js [hunk 12]
```diff
@@ -1942,6 +2005,58 @@ describe('Facebook Pixel', function() {
...
+          'Purchase',
+          {
+            content_ids: [
+              '507f1f77bcf86cd799439011',
+              '505bd76785ebb509fc183733'
+            ],
+            content_type: ['myCustomType'],
+            currency: 'USD',
+            value: '0.50',
+            content_name: 'my product',
```
Comment (brennan) [inline integrations/facebook-pixel/test/index.test.js]: I don't think we need to map `content_name` here.

File: integrations/facebook-pixel/test/index.test.js [hunk 13]
```diff
@@ -1992,6 +2107,25 @@ describe('Facebook Pixel', function() {
...
+          query: 'yo',
+          currency: 'USD',
+          category: 'Cars',
+          revenue: 30
+        });
+        analytics.called(window.fbq, 'trackSingle', options.pixelId, 'Search', {
+          search_string: 'yo',
+          content_category: 'Cars',
+          content_ids: ['507f1f77bcf86cd799439011'],
+          contents: [],
```
Comment (brennan) [inline integrations/facebook-pixel/test/index.test.js]: Why is the contents array empty?

**Review by brennan — CHANGES_REQUESTED** [review]
Left some comments - thanks!

File: integrations/facebook-pixel/test/index.test.js (hunk 8 above)
Comment (vjnathe-webonise) [reply integrations/facebook-pixel/test/index.test.js:1405]: Got that.

File: integrations/facebook-pixel/test/index.test.js (hunk 10 above)
Comment (vjnathe-webonise) [reply integrations/facebook-pixel/test/index.test.js:1842]: Got that.

File: integrations/facebook-pixel/test/index.test.js (hunk 12 above)
Comment (vjnathe-webonise) [reply integrations/facebook-pixel/test/index.test.js]: Got that, **content_name** is missing field but we only need to map it if its corresponding field present user specs.
Also changed **total** to **revenue** in analytics mapping

File: integrations/facebook-pixel/test/index.test.js (hunk 13 above)
Comment (vjnathe-webonise) [reply integrations/facebook-pixel/test/index.test.js]: Ohh, i missed this, i was under impression that contents will purely derived from products array.

Comment (vjnathe-webonise) [reply integrations/facebook-pixel/test/index.test.js]: @Brennan removing this test case completely as there is no corresponding specs present here: https://segment.com/docs/spec/ecommerce/v2/#products-searched for missing properties.

**Comment by vjnathe-webonise** [comment]
@brennan I resolved the PR comments, also removed test case for **Products Searched** as there are no mapping specs for missing properties.

**Review by CarlosMecha — APPROVED** [review]
Tests are good! :+1:

**Comment by vjnathe-webonise** [comment]
@CarlosMecha @brennan I have added the integration code for missing properties for facebook standard events.

File: integrations/facebook-pixel/lib/index.js [hunk 14]
```diff
@@ -198,15 +198,22 @@ FacebookPixel.prototype.productListViewed = function(track) {
   var self = this;
   var contentType;
   var contentIds = [];
+  var contents = [];
   var products = track.products();
   var customProperties = this.buildPayload(track, true);
 
   // First, check to see if a products array with productIds has been defined.
   if (Array.isArray(products)) {
     products.forEach(function(product) {
```
Comment (CarlosMecha) [inline integrations/facebook-pixel/lib/index.js:206]: I know this wasn't introduced in this PR, but could we change the `forEach` to a traditional `for...loop` for the products array?

File: integrations/facebook-pixel/lib/index.js [hunk 15]
```diff
@@ -359,22 +385,25 @@ FacebookPixel.prototype.orderCompleted = function(track) {
...
-  );
-
   var revenue = formatRevenue(track.revenue());
 
   // Order completed doesn't have a top-level category spec'd.
   // Let's default to the category of the first product. - @gabriel
   var contentType = this.getContentType(track, ['product']);
+  var contentIds = [];
+  var contents = [];
+  each(function(product) {
```
Comment (CarlosMecha) [inline integrations/facebook-pixel/lib/index.js]: Could we change this `each` to a traditional `for...loop` for the products array?

**Review by CarlosMecha — CHANGES_REQUESTED** [review]
Left a couple of comments, but overall :100:

**Comment by vjnathe-webonise** [comment]
@CarlosMecha Changed the iterator as per suggestion, please have a look.

**Review by CarlosMecha — APPROVED** [review]

**Comment by vjnathe-webonise** [comment]
Closing this PR as it's merged to upstream.

**Pull request closed without merging**